subsystem/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): L. Kuklinek
+
+//! General framework for working with subsystems
+
+use std::future::Future;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio::task;
+
+use logging::log;
+
+/// Top-level subsystem manager.
+///
+/// An application is composed of a number of long-lived subsystems. The [Manager] type starts
+/// and manages the life cycle of the subsystems. Whenever a subsystem exits, all other subsystems
+/// are requested to terminate and the manager is shut down.
+pub struct Manager {
+    // Manager name
+    name: &'static str,
+
+    // Order all subsystems to shutdown
+    shutdown_request_tx: broadcast::Sender<()>,
+    shutdown_request_rx: broadcast::Receiver<()>,
+
+    // Signal a subsystem is shutting down. This is taken as an instruction for all subsystems to
+    // shut down. Shutdown completion is detected by all senders having closed this channel.
+    shutting_down_tx: mpsc::Sender<()>,
+    shutting_down_rx: mpsc::Receiver<()>,
```

Comment:
It seems all we do with `shutting_down_request_rx` is drop it in main. Why do we need to keep it at all? What would happen if we were to not store this receiver handle in the `Manager`?

Reply:
Maybe it could be removed, let me check. Dropping this immediately after channel creation would close the channel which is taken as a signal that the system has terminated, so not sure if this would trigger a premature shutdown.

Reply:
This seems to have some interesting effects. If the main task does not hang on to its end of the channel, the behaviour when there are no subsystem changes. Before it would just wait because it is waiting for a subsystem to signal termination. If there is no such subsystem there is nobody to tell the whole system to stop. With some code changes, we can make it so that the main thread just exits immediately if there are no subsystems, which seems like a cleaner solution.

Reply:
I have now implemented the cleaner shutdown semantics with respect to no subsystems scenario. Also the Ctrl-C handler now has to be explicitly enabled so it does not get in the way in tests etc.